packages/models-library/src/models_library/settings/base.py

```diff
@@ -0,0 +1,35 @@
+from contextlib import suppress
+from typing import List, Tuple
+
+from pydantic import BaseSettings, Extra, SecretStr, ValidationError
+
+
+class BaseCustomSettings(BaseSettings):
+    class Config:
+        case_sensitive = False
+        extra = Extra.forbid
+        allow_mutation = False
+        frozen = True
+        validate_all = True
+        json_encoders = {SecretStr: lambda v: v.get_secret_value()}
+
+    @classmethod
+    def set_defaults_with_default_constructors(
+        cls, default_fields: List[Tuple[str, "BaseCustomSettings"]]
+    ):
+        # This funcion can set defaults on fields that are BaseSettings as well
+        # It is used in control construction of defaults.
+        # Pydantic offers a defaults_factory but it is executed upon creation of the Settings **class**
+        # which is too early for our purpose. Instead, we want to create the defaults just
+        # before the settings instance is constructed
```

Comment:
is this solving the issue of accessing the env on creation?
not sure I understand exactly what goes on in the next for loop.

Reply:
yes.
the problem is that we cannot create defaults for the sub-settings fields (i.e. those fields that are Settings themselves as well). Even using the ``default_factory``, the actual default is instantiated when the **class** (not the **object**) is created. I wanted to avoid globals so I created a new "default_factory" where we instanciate the default just before creating the instance. That implies that defaults might change for every instance but in this particular case settings are scoped with the lifetime of an app so it is not a problem.